dingo/gw/noise/synthetic_noise/generate_dataset.py

```diff
@@ -0,0 +1,117 @@
+from dingo.gw.noise.synthetic_noise.utils import reconstruct_psds_from_parameters
+
+
+def parse_args():
+    parser = argparse.ArgumentParser(
+        formatter_class=argparse.RawDescriptionHelpFormatter,
+        description=textwrap.dedent(
+            """\
+        Generate a synthetic noise ASD dataset from an existing dataset of real ASDs.
+        """
+        ),
+    )
+    parser.add_argument(
+        "--asd_dataset",
+        type=str,
+        required=True,
+        help="Path to existing ASD dataset to be parameterized and re-sampled",
+    )
+    parser.add_argument(
+        "--settings_file",
+        type=str,
+        required=True,
+        help="YAML file containing database settings",
+    )
+    parser.add_argument(
+        "--num_processes",
+        type=int,
+        default=1,
+        help="Number of processes to use in pool for parallel parameterization",
+    )
+    parser.add_argument(
+        "--out_file",
+        type=str,
+        default="synthetic_asd_dataset.hdf5",
+        help="Name of file for storing dataset.",
+    )
+    parser.add_argument("--verbose", action="store_true")
+
+    return parser.parse_args()
+
+
+def generate_dataset(real_dataset, settings: Dict, num_processes: int, verbose: bool):
+    parameters_dict = parameterize_asd_dataset(
+        real_dataset,
+        settings["parameterization_settings"],
+        num_processes,
+        verbose=verbose,
+    )
+
+    synthetic_dataset_dict = copy.deepcopy(real_dataset.to_dictionary())
+    synthetic_dataset_dict["settings"]["parameterization_settings"] = settings[
+        "parameterization_settings"
+    ]
+
+    sampling_settings = settings.get("sampling_settings", None)
+    if sampling_settings:
+        kde = KDE(parameters_dict, sampling_settings)
+        kde.fit()
+        rescaling_params = None
+        if "rescaling_psd_paths" in sampling_settings:
+            rescaling_params = get_rescaling_params(
+                sampling_settings["rescaling_psd_paths"],
+                settings["parameterization_settings"],
+            )
+        parameters_dict = kde.sample(rescaling_params)
+        synthetic_dataset_dict["settings"]["sampling_settings"] = settings["sampling_settings"]
+
+    asds_dict = {}
+    for det, params in parameters_dict.items():
+        psds = reconstruct_psds_from_parameters(
+            params,
+            real_dataset.domain,
+            settings["parameterization_settings"],
+        )
+
+        asds_dict[det] = np.sqrt(
+            psds[:, real_dataset.domain.min_idx : real_dataset.domain.max_idx + 1]
+        )
+
+    synthetic_dataset_dict["parameters"] = parameters_dict
+    synthetic_dataset_dict["asds"] = asds_dict
+    # TODO: should I make the "data_keys" an optional argument to the ASDDataset class? Such that the parameters
+    # can also be stored in this way
+    return DingoDataset(
+        dictionary=synthetic_dataset_dict, data_keys=["asds", "gps_times", "parameters"]
+    )
```

Comment:
Is there a reason you don't use an `ASDDataset` here?

Reply:
Also this preserves the GPS times from the ASDDataset that you started from, even though the individual synthetic ASDs have nothing to do with the original ones. It would be better to drop the "gps_times" key from the dict, and make sure the `ASDDataset` is robust to this key missing.

Reply:
I cannot overwrite the `data_keys` of ASDDataset. Using ASDDataset wouldn't allow me to save `parameters`, see the comment in the code. Regarding the gps_times, this is intentional, since I thought it would be good to know data from which time window I've accessed to created the synthetic dataset, in particular if the PSDs were only parameterised and not re-sampled

Reply:
Ah I see, you are saving the parameters as well. I guess when this gets loaded later by an ASDDataset, it just ignores the parameters. The only issue I foresee is that the `dataset_type` would not be set to `"asd_dataset"`, which would affect the output of `dingo_ls`. You could alternatively add `"parameters"` to the ASDDatset `data_keys`. (To avoid confusion, it might be better to call it `"asd_parametrizations"` or something like that.

Reply:
If it's not too much work, I would suggest it.